tests/Unit/NumericalAlgorithms/LinearOperators/Test_WeakDivergence.cpp

```diff
@@ -0,0 +1,163 @@
+#include "NumericalAlgorithms/LinearOperators/WeakDivergence.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace {
+struct Var1 : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct Var2 : db::SimpleTag {
+  using type = tnsr::i<DataVector, Dim>;
+};
+
+template <size_t Dim>
+void test_weak_divergence_random_jacobian(const Mesh<Dim>& mesh) {
+  CAPTURE(Dim);
+  CAPTURE(mesh);
+  CAPTURE(mesh.quadrature(0));
+  std::uniform_real_distribution<double> dist(-1.0, 2.3);
+  MAKE_GENERATOR(gen);
+  using flux_tags =
+      tmpl::list<Tags::Flux<Var1, tmpl::size_t<Dim>, Frame::Inertial>,
+                 Tags::Flux<Var2<Dim>, tmpl::size_t<Dim>, Frame::Inertial>>;
+  using div_tags = tmpl::list<
+      Tags::div<Tags::Flux<Var1, tmpl::size_t<Dim>, Frame::Inertial>>,
+      Tags::div<Tags::Flux<Var2<Dim>, tmpl::size_t<Dim>, Frame::Inertial>>>;
+
+  tnsr::I<DataVector, Dim, Frame::Inertial> inertial_coords{
+      mesh.number_of_grid_points()};
+  Jacobian<DataVector, Dim, Frame::Logical, Frame::Inertial> jacobian{
+      mesh.number_of_grid_points()};
+
+  fill_with_random_values(make_not_null(&inertial_coords), make_not_null(&gen),
+                          make_not_null(&dist));
+  fill_with_random_values(make_not_null(&jacobian), make_not_null(&gen),
+                          make_not_null(&dist));
+
+  InverseJacobian<DataVector, Dim, Frame::Logical, Frame::Inertial>
+      det_jac_times_inverse_jacobian{};
+
+  dg::metric_identity_jacobian(make_not_null(&det_jac_times_inverse_jacobian),
+                               mesh, inertial_coords, jacobian);
+  // Generate constant fluxes that aren't all the same.
+  Variables<flux_tags> fluxes{mesh.number_of_grid_points(), 2.0};
+  tmpl::for_each<flux_tags>([&fluxes](auto tag_v) noexcept {
+    using tag = tmpl::type_from<decltype(tag_v)>;
+    auto& flux = get<tag>(fluxes);
+    for (size_t storage_index = 0; storage_index < flux.size();
+         ++storage_index) {
+      flux[storage_index] = storage_index + 3.0;
+    }
+  });
+
+  Variables<div_tags> divergence_result{mesh.number_of_grid_points()};
+
+  weak_divergence(make_not_null(&divergence_result), fluxes, mesh,
+                  det_jac_times_inverse_jacobian);
+
+  Approx local_approx = Approx::custom().epsilon(5.0e-12).scale(1.);
+  const Variables<div_tags> expected_divergence_result{
+      mesh.number_of_grid_points(), 0.0};
+  tmpl::for_each<div_tags>([&divergence_result, &expected_divergence_result,
+                            &local_approx, &mesh](auto tag_v) {
+    using tag = tmpl::type_from<decltype(tag_v)>;
+    if (mesh.quadrature(0) == Spectral::Quadrature::GaussLobatto) {
+      // Only the interior points are zero. Points on the boundary are non-zero
+      // because they need the flux lifted to be zero.
```

Comment:
I don't understand what you're saying here (not even enough to ask a specific clarifying question). Can you expand?

Can you explain, in a github response for now, why the result of the divergence should have any dependence on the concept of flux-lifting? This is deeply upsetting if true...

Reply:
Well I think you'll have to be deeply upset then ;)

What happens is you do an integration by-parts, which gives you a boundary contribution and a volume contribution. It is the _sum_ of those that is zero. With GL points the surface integral only contributes to the grid points on the surface, and thus it is only the interior points that are zero. With Gauss points the surface integral contributes to "all" points effectively. Does that help? If not, I can type up some math in a comment, but the short of it is:

P d_i F^i = 0

but for weak form we have

F^i d_i p != P d_i F^i

instead,

P d_i F^i = - F^i d_i P + surface terms = 0

Does that make sense?

Reply:
👍 Nils and had a conversation to clear up my misunderstandings .. Nils will add some documentation to clarify what is being tested and why